### PR TITLE
Add Async Support 

### PR DESCRIPTION
**Pull request by theGowda**
Added async support. Uses asyncio.open_connection which provied StreamReader and Writer

### CODE REVIEW

**Comment by theGowda** [comment]
@luqasz any thoughts on this?

File: README.rst [hunk 1]
```diff
@@ -3,3 +3,14 @@ Documentation
 
 Documentation resides over at
 `readthedocs <https://librouteros.readthedocs.io/>`_
+
+Async Usage
```
Comment (luqasz) [inline README.rst:7]: Move it to docs.

File: librouteros/query.py [hunk 2]
```diff
@@ -71,3 +71,24 @@ def Or(left: QueryGen, right: QueryGen, *rest: QueryGen) -> QueryGen:
...
+        self.query = tuple(chain.from_iterable(args))
+        return self
+
+    def __aiter__(self):
+        keys = ",".join(str(key) for key in self.keys)
+        keys = f"=.proplist={keys}"
+        cmd = str(self.path.join("print"))
+        return self.api.rawCmd(cmd, keys, *self.query)
+
+    def __iter__(self):
```
Comment (luqasz) [inline librouteros/query.py:93]: Since it's async only, no need for `__iter__` method at all.

**Review by luqasz — CHANGES_REQUESTED** [review]

**Comment by luqasz** [comment]
@theGowda Do you want to add tests ?

File: librouteros/query.py (hunk 2 above)
Comment (theGowda) [reply librouteros/query.py:93]: It is prevent anyone from using list expression on it. Just throws an error.

**Comment by theGowda** [comment]
> @theGowda Do you want to add tests ?

I have never written async tests. Basically have to copy everything with async await. 
I'll create another folder inside test for async tests

File: librouteros/query.py (hunk 2 above)
Comment (luqasz) [reply librouteros/query.py:93]: If it's not present python should throw `AttributeError` as far as I remember.

File: librouteros/query.py [hunk 3]
```diff
@@ -71,3 +71,24 @@ def Or(left: QueryGen, right: QueryGen, *rest: QueryGen) -> QueryGen:
...
+
+    def where(self, *args: str):
+        self.query = tuple(chain.from_iterable(args))
+        return self
+
+    def __aiter__(self):
+        keys = ",".join(str(key) for key in self.keys)
+        keys = f"=.proplist={keys}"
+        cmd = str(self.path.join("print"))
+        return self.api.rawCmd(cmd, keys, *self.query)
```
Comment (luqasz) [inline librouteros/query.py:91]:
```suggestion
        words = tuple(self.query)
        if len(self.keys) > 0:
            keys = ",".join(str(key) for key in self.keys)
            keys = f"=.proplist={keys}"
            words = (keys,) + words
        return iter(self.api.rawCmd(cmd, *words))
```
From https://github.com/luqasz/librouteros/commit/1ee5479df767e59923f35e2c2ed124f17d0bf186

File: librouteros/query.py (hunk 2 above)
Comment (theGowda) [reply librouteros/query.py:99]: changed to `AttributeError `

**Comment by luqasz** [comment]
@theGowda you can run tests locally.

**Comment by theGowda** [comment]
> @theGowda you can run tests locally.

Ya the integration tests passed.
I am updating the unit tests

File: tests/integration/async/test_general.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                password="",
+                login_method=method,
+            )
+            break
+        except (LibRouterosError, socket.error, socket.timeout):
+            sleep(1)
+
+    # data = api("/system/identity/print")
+    result = [r async for r in api("/system/identity/print")]
+    assert result[0]["name"] == "Mikrotik-Sachin"
```
Comment (luqasz) [inline tests/integration/async/test_general.py:28]: `Mikrotik-Sachin` where does this come from ?

**Comment by luqasz** [comment]
> > @theGowda you can run tests locally.
> 
> Ya the integration tests passed. I am updating the unit tests

https://github.com/luqasz/librouteros/actions/runs/12126034808/job/33859566156

File: tests/integration/async/test_general.py (hunk 4 above)
Comment (theGowda) [reply tests/integration/async/test_general.py:28]: Sorry about that. Was testing something.

**Comment by theGowda** [comment]
> > > @theGowda you can run tests locally.
> > 
> > 
> > Ya the integration tests passed. I am updating the unit tests
> 
> https://github.com/luqasz/librouteros/actions/runs/12126034808/job/33859566156

I had to install pytest-asyncio as dev dependency for the async tests. Is that causing it? Because tests seem to pass locally.
I'll check

**Comment by luqasz** [comment]
```
pyproject.toml changed significantly since poetry.lock was last generated. Run `poetry lock [--no-update]` to fix the lock file.
```
Everything is written in actions output.

**Comment by theGowda** [comment]
> ```
> pyproject.toml changed significantly since poetry.lock was last generated. Run `poetry lock [--no-update]` to fix the lock file.
> ```
> 
> Everything is written in actions output.

i'll complete the unit tests and push a commit

File: README.rst (hunk 1 above)
Comment (theGowda) [reply README.rst]: Done

File: tests/integration/async/test_general.py (hunk 4 above)
Comment (theGowda) [reply tests/integration/async/test_general.py]: Done

**Comment by luqasz** [comment]
Async integration tests are failing. Looks like it can't connect. You can [download](https://github.com/luqasz/librouteros/blob/main/.github/workflows/test.yml#L17) docker images with routeros vms.

**Comment by theGowda** [comment]
> Async integration tests are failing. Looks like it can't connect. You can [download](https://github.com/luqasz/librouteros/blob/main/.github/workflows/test.yml#L17) docker images with routeros vms.

But I dont understand why it is failing. I ran integration tests locally using  qcow2 images as suggested in the readme file.
Worked fine.

**Comment by luqasz** [comment]
Maybe increase timeout in `async_connect()`

**Comment by theGowda** [comment]
> Maybe increase timeout in `async_connect()`

changed it to 60 for now in tests

**Comment by theGowda** [comment]
@luqasz can you please re-run the failed tests. it will pass.

**Comment by theGowda** [comment]
> ```
> /__w/librouteros/librouteros/.venv/lib/python3.8/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
> The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"
> ```
> 
> Pytest throws this warning both ni unit and integration tests.

That's fine. Just a warning. I don't think it'll affect the functioning

**Comment by luqasz** [comment]
OK. Final review.

**Comment by luqasz** [comment]
@theGowda can you please move `integration/async` into `integration ? IMO it's unnecessary to duplicate fixture code. Adding async tests into existing files will be fine.

**Comment by theGowda** [comment]
> @theGowda can you please move `integration/async` into `integration ? IMO it's unnecessary to duplicate fixture code. Adding async tests into existing files will be fine.



> @theGowda can you please move `integration/async` into `integration ? IMO it's unnecessary to duplicate fixture code. Adding async tests into existing files will be fine.

Okay. I'll move the tests

**Comment by theGowda** [comment]
@luqasz i think you sould stop the test for python 3.12. its been running for 10mins now

**Comment by luqasz** [comment]
Seems that on `3.12` tests timeout but only for routeros `6.33.3`. Does those run locally ?

**Comment by theGowda** [comment]
> Seems that on `3.12` tests timeout but only for routeros `6.33.3`. Does those run locally ?

Yes. Github action succeeded in my fork.
Screenshot 2024-12-08 at 8 16 03 PM

**Review by luqasz — APPROVED** [review]

**Pull request merged**